File: backend/src/main/java/top/vuhe/common/ApiResponse.java
package top.vuhe.common;

import lombok.Getter;
import top.vuhe.common.exception.ExceptionEnum;
import top.vuhe.common.exception.SystemProcessingException;

/**
 * 通用返回类
 *
 * @author zhuhe
 */
@Getter
public class ApiResponse<T> {
    private int code;
    private String message;
    private String dataName;
    private T data;

    private ApiResponse() {
    }

    private ApiResponse(int code, String message, String name, T data) {
        this.code = code;
        this.message = message;
        this.dataName = name;
        this.data = data;
    }

    public static <T> ApiResponse<T> of(Integer code, String message, String name, T data) {
        return new ApiResponse<>(code, message, name, data);
    }

    public static <T> ApiResponse<T> of(Integer code, String message) {
        return new ApiResponse<>(code, message, "", null);
    }

    public static <T> ApiResponse<T> ofSuccess() {
        return of(200, "success");
    }

    public static <T> ApiResponse<T> ofSuccessWithDate(String name, T date) {
        return of(200, "success", name, date);
    }

    public static <T> ApiResponse<T> ofException(SystemProcessingException e) {
        return of(e.getCode(), e.getMessage());
    }

    public static <T> ApiResponse<T> ofErrorEnum(ExceptionEnum e) {
        return of(e.getCode(), e.getMessage());
    }

}

File: backend-kt/src/main/kotlin/top/vuhe/sw/common/ApiResponse.kt
package top.vuhe.sw.common

import com.fasterxml.jackson.core.JsonGenerator
import com.fasterxml.jackson.databind.SerializerProvider
import com.fasterxml.jackson.databind.annotation.JsonSerialize
import com.fasterxml.jackson.databind.ser.std.StdSerializer
import io.swagger.annotations.ApiModel
import io.swagger.annotations.ApiModelProperty
import top.vuhe.sw.common.exception.ExceptionEnum
import top.vuhe.sw.common.exception.SystemProcessingException

/**
 * 通用返回类
 *
 * @author zhuhe
 */
@ApiModel(value = "通用返回数据")
@JsonSerialize(using = ApiResponseSerializer::class)
class ApiResponse<T> private constructor(
    @field:ApiModelProperty(value = "状态代码", readOnly = true)
    val code: Int,
    @field:ApiModelProperty(value = "状态信息", readOnly = true)
    val message: String,
    @field:ApiModelProperty(value = "数据", readOnly = true, allowEmptyValue = true)
    val data: T?
) {
    companion object {
        fun <T> of(code: Int, message: String, data: T?): ApiResponse<T> {
            return ApiResponse(code, message, data)
        }

        fun of(code: Int, message: String): ApiResponse<*> {
            return of<Any>(code, message, null)
        }

        fun ofSuccess(): ApiResponse<*> {
            return of(200, "success")
        }

        fun <T> ofSuccessWithDate(date: T): ApiResponse<T> {
            return of(200, "success", date)
        }

        fun ofException(e: SystemProcessingException): ApiResponse<*> {
            return of(e.code, e.msg)
        }

        fun ofErrorEnum(e: ExceptionEnum): ApiResponse<*> {
            return of(e.code, e.message)
        }
    }
}

/**
 * 序列化 通用返回类
 *
 * @author zhuhe
 */
class ApiResponseSerializer : StdSerializer<ApiResponse<*>> {

    constructor() : this(null)

    constructor(t: Class<ApiResponse<*>>?) : super(t)

    override fun serialize(
        value: ApiResponse<*>,
        gen: JsonGenerator,
        provider: SerializerProvider?
    ) {
        gen.writeStartObject()
        gen.writeNumberField("code", value.code)
        gen.writeStringField("message", value.message)
        if (value.data != null) {
            gen.writeObjectField("data", value.data)
        }
        gen.writeEndObject()
    }

}